### PR TITLE
Don't use strcasecmp and strncasecmp for IA5 strings

### PR DESCRIPTION
**Pull request by mattcaswell**
The functions strcasecmp() and strncasecmp() will use locale specific rules
when performing comparison. This could cause some problems in certain
locales. For example in the Turkish locale an 'I' character is not the
uppercase version of 'i'. However IA5 strings should not use locale specific
rules, i.e. for an IA5 string 'I' is uppercase 'i' even if using the
Turkish locale.

This fixes a bug in name constraints checking reported by Thomas Porin
(NCCGroup).

This is not considered a security issue because it would require both a
Turkish locale (or other locale with similar issues) and malfeasance by
a trusted name-constrained CA for a certificate to pass name constraints
in error. Failure to match permitted subtrees is a bug, not a vulnerability.

### CODE REVIEW

**Comment by vdukhovni** [comment]
Code looks good to me.  The commit message could say that only does the CA have to misbehave, but it needs to be a CA constrained with excluded subtrees.  I've never seen one of those in the wild...

**Comment by tmshort** [comment]
Not asking for any changes, but I'm just going to add a note that this is not compatible with the EBCDIC character set.

**Comment by tmshort** [comment]
Not asking for any changes, but I'm just going to add a note that this is not compatible with the EBCDIC character set.

**Comment by mattcaswell** [comment]
> Not asking for any changes, but I'm just going to add a note that this is not compatible with the EBCDIC character set.

I'm not sure what you mean by that. ia5 is essentially an ASCII based system so it should not be processed as EBCDIC even on an EBCDIC based system. Where do you see a problem?

**Comment by tmshort** [comment]
Yup... character constants are not used, you use hex numbers... so... never mind. EBCDIC considerations have been taken!

**Comment by tmshort** [comment]
(Same with #4570)

File: crypto/x509v3/v3_ncons.c [hunk 1]
```diff
@@ -64,6 +63,45 @@ ASN1_SEQUENCE(NAME_CONSTRAINTS) = {
 IMPLEMENT_ASN1_ALLOC_FUNCTIONS(GENERAL_SUBTREE)
 IMPLEMENT_ASN1_ALLOC_FUNCTIONS(NAME_CONSTRAINTS)
 
+/*
+ * We cannot use strncasecmp here because that applies locale specific rules.
+ * For example in Turkish 'I' is not the uppercase character for 'i'. We need to
+ * do a simple ASCII case comparison ignoring the locale.
+ */
```
Comment (richsalz) [inline crypto/x509v3/v3_ncons.c]: Please add a comment that says "that is why we use numeric constants, below" or something similar.

Comment (richsalz) [reply crypto/x509v3/v3_ncons.c]: And same comment in #4570 too.

**Comment by dot-asm** [comment]
On related note, master has locale-neutral ossl_to{lower|upper}. [Which also aims to provide EBCDIC-friendly interface. "Aims" in last sentence means rather "intention to figure it out as soon as access to z/OS is provided".]

File: crypto/x509v3/v3_ncons.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/x509v3/v3_ncons.c]: Done

**Comment by mattcaswell** [comment]
I updated this to address the comment from @richsalz. Also updated the commit message as requested by @vdukhovni. Please reconfirm.

**Comment by mattcaswell** [comment]
Fixed a minor typo in the reporter's name in the commit message. I will assume approvals still apply.

**Comment by levitte** [comment]
Yup, here and in the  1.0.2 PR

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**